Add CCTP TokenBridge support

gitmp01 opened the pull request:
### Description

<!--
What's included in this PR?
-->
- Add CCTP TokenBridge contracts

### Drive-by changes

<!--
Are there any minor or drive-by changes also included?
-->

### Related issues

<!--
- Fixes #[issue number here]
-->

### Backward compatibility

<!--
Are these changes backward compatible? Are there any infrastructure implications, e.g. changes that would prohibit deploying older commits using this infra tooling?

Yes/No
-->

### Testing

<!--
What kind of testing have these changes undergone?

None/Manual/Unit Tests
-->

gitmp01 committed:
chore: add `CCTP` contracts

gitmp01 committed:
chore: add CCTP CCIP-read service

gitmp01 committed:
chore: remove

gitmp01 committed:
chore(scripts): add `TokenBridgeCctp.s.sol` script

gitmp01 committed:
forge install: memview-sol

latest-solidity

gitmp01 committed:
feat: handle CCTP v2 standard transfer

gitmp01 committed:
fix(CctpHook): rm duplicate import

gitmp01 committed:
refactor(CctpIsm.sol): factor out nonce retrieval + access once `usedNonces`

gitmp01 committed:
chore(CCTPServiceAbi.ts): add comment

gitmp01 committed:
refactor(CCTPAttestationService.ts): pull magic constants out

gitmp01 committed:
chore(CCTPService.ts): point out location in comment

gitmp01 committed:
chore(CctpHook.sol): adjust links to source

gitmp01 committed:
fix: use strong typing for CCTP services

gitmp01 committed:
fix(CCTPService): use `const`

gitmp01 committed:
fix(HyperlaneService.ts): reduce limit to 1

gitmp01 committed:
chore: use `GasRouter` & refactor

gitmp01 committed:
chore(TokenBridgeCctp.sol): pass through `scale` param

yorhodes commented on solidity/contracts/isms/CctpIsm.sol:
```diff
@@ -0,0 +1,100 @@
+    using CctpMessageV2 for bytes29;
+
+    string[] urls;
+    IMailbox mailbox;
+    IMessageTransmitter messageTransmitter;
```
these should be immutables

yorhodes commented on solidity/contracts/isms/CctpIsm.sol:
```diff
@@ -0,0 +1,100 @@
+        bytes29 cctpMessage,
+        uint32 cctpVersion
+    ) internal pure returns (bytes32 nonce) {
+        if (cctpVersion > 1) revert UnsupportedCCTPVersion(cctpVersion);
+
+        if (cctpVersion == 0) {
```
can you reuse the constants from `TokenBridgeCCTP`?

yorhodes replied:
in fact I think we should have separate V1 and V2 implementations

yorhodes commented on solidity/contracts/isms/CctpIsm.sol:
```diff
@@ -0,0 +1,100 @@
+            // CCTP v1
+            uint64 nonceV1 = cctpMessage._getNonceV1();
+            uint32 sourceDomain = cctpMessage._getSourceDomain();
+            nonce = keccak256(abi.encodePacked(sourceDomain, nonceV1));
+        } else {
+            // CCTP v2
+            nonce = cctpMessage._getNonce();
```
nit: prefer explicit returns

yorhodes commented on solidity/contracts/client/GasRouter.sol:
```diff
@@ -59,6 +59,13 @@ abstract contract GasRouter is Router {
     function quoteGasPayment(
         uint32 _destinationDomain
     ) external view virtual returns (uint256) {
+        return _quoteGasPayment(_destinationDomain);
+    }
+
+    /// @dev Lets derived contracts access the quote
+    function _quoteGasPayment(
+        uint32 _destinationDomain
+    ) internal view virtual returns (uint256) {
```
alternatively you could make the `quoteGasPayment` public

gitmp01 replied:
Is there a rule of thumb for this? because I thought that option as well, but it was hard to choose :sweat_smile:

yorhodes commented on solidity/contracts/token/extensions/TokenBridgeCctp.sol:
```diff
@@ -0,0 +1,138 @@
+    ) internal {
+        wrappedToken.approve(address(tokenMessenger), _amount);
+        uint32 circleDomain = hyperlaneDomainToCircleDomain[_destination];
+        uint32 version = tokenMessenger.messageBodyVersion();
+
+        bytes32 router = _mustHaveRemoteRouter(_destination);
+
+        if (version == CCTP_VERSION_1) {
+            tokenMessenger.depositForBurn(
+                _amount,
+                circleDomain,
+                router,
+                address(wrappedToken)
+            );
+        } else if (version == CCTP_VERSION_2) {
+            ITokenMessengerV2(address(tokenMessenger)).depositForBurn(
+                _amount,
+                circleDomain,
+                router,
+                address(wrappedToken),
+                bytes32(0),
+                CCTP_V2_DEFAULT_MAX_FEE,
+                CCTP_V2_DEFAULT_MIN_FINALITY_THRESHOLD
+            );
+        } else {
+            revert UnsupportedCCTPVersion(version);
+        }
```
tokenMessenger is immutable so each deployment of this will only ever use 1 branch
in this case it makes sense to have 2 separate implementations that assert the version matches at deployment time and does not need to have branching logic here

yorhodes replied:
also gets around the interface casting issue

gitmp01 replied:
Makes sense, I'll prob be able to do it in the coming days, but can't test it until I'm back

yorhodes commented on typescript/ccip-server/src/server.ts:
can you remove the OP stack related changes from this PR?

gitmp01 replied:
There are things that are useful from the OPStack one, like the ccip-read hyperlane service and others I guess, do you want me to do it anyway?

yorhodes commented on solidity/contracts/libs/CctpMessageV2.sol:
```diff
@@ -0,0 +1,137 @@
+        return _message.index(NONCE_INDEX, 32);
+    }
+
+    // @dev Copied from _nonce() of Message.sol of the CCTP repo in order to support
+    // CCTP v1 messages
+    function _getNonceV1(bytes29 _message) internal pure returns (uint64) {
+        return uint64(_message.indexUint(NONCE_INDEX, 8));
+    }
```
I dont like the idea of making modifications to vendored code
can we import both V1 and V2 libraries directly as is instead of trying to make our own merged version?

gitmp01 replied:
Sure

nambrot commented on solidity/contracts/isms/CctpIsm.sol:
```diff
@@ -0,0 +1,100 @@
+import {TypedMemView} from "@memview-sol/contracts/TypedMemView.sol";
+import {IMessageTransmitter} from "../interfaces/cctp/IMessageTransmitter.sol";
+
+contract CctpIsm is AbstractCcipReadIsm {
```
Wondering if this should just be part of the `ITokenBridge`? To make managing artifacts easier? I.e. the ISM of `TokenBridgeCctp` should always be this ism no?

gitmp01 replied:
Yeah, would like me to deploy it within the constructor?

nambrot replied:
Yeah, imo it should even just be part of the same contract and then `interchainSecurityModule` can just return `address(this)` That we don't have to track separate contracts?

nambrot committed:
chore: vendor TypedMemView

Add TypedMemView.sol

chore: forge rm `memview-sol`

gitmp01 committed:
chore: divide into CCTP V1 & CCTP V2, include CctpIsm into TokenBridgeCctp

chore: split V1 & V2 into two different contracts + include ISM within the router

gitmp01 committed:
chore(TokenBridgeCctpScript): adapt according to v1 & v2 changes

gitmp01 committed:
chore: use `routers` function

gitmp01 committed:
chore(.gitignore): ignore `*.env`

gitmp01 committed:
chore: move infinite approval into the constructor

gitmp01 committed:
chore: add test suite

gitmp01 committed:
chore: rm setHook

larryob committed:
feat: Decode internal calls in forge tests (#6224)

### Description

By default we decode internal calls in our foundry tests. This is great
for debugging! See these [Foundry v1.0 release
notes](https://www.paradigm.xyz/2025/02/announcing-foundry-v1-0) for
more.

larryob committed:
fix(ICA): Remove ica from commitment  (#6211)

### Description

The commitment is simply `hash(salt, calls)`. 

### Drive by
- Remove redundant checks on revealed data
- Move check on existing commitment into ICA.

---------

Co-authored-by: Yorke Rhodes <yorke@hyperlane.xyz>

gitmp01 committed:
feat: add OpStack L2 to L1 ITokenBridge (#5645)

### Description

1. Adds `ValueTransferBridgeNative` router abstracting the transfer of
ETH from L2 to L1
2. Adds `OPValueTransferBridgeNative` offering an implementation for 1.
on Optimism
 3. Adds hooks and required ISM for 2.
4. Adds an `OPStackService` to the CCIP-read server handling the
CCIP-read requests

<!--
What's included in this PR?
-->

### Drive-by changes

<!--
Are there any minor or drive-by changes also included?
-->

### Related issues

<!--
- Fixes #[issue number here]
-->

### Backward compatibility

<!--
Are these changes backward compatible? Are there any infrastructure
implications, e.g. changes that would prohibit deploying older commits
using this infra tooling?

Yes/No
-->

### Testing

<!--
What kind of testing have these changes undergone?

None/Manual/Unit Tests
-->

larryob committed:
feat(ICA): Allow multiple commitments for a given ica (#6221)

### Description
You can have multiple pending commitments for a given ica. The
commitments are completely independent of each other.

---------

Co-authored-by: Nam Chu Hoai <nambrot@googlemail.com>

gitmp01 committed:
fix(TokenBridgeCctp.sol): use public function in GasRouter

gitmp01 committed:
fix(TokenBridgeCctp.t.sol): rm duplicate import

gitmp01 committed:
fix(TokenBridgeCctp.t.sol): test name

gitmp01 committed:
chore(MockCircleMessageTransmitter): just return the version

gitmp01 committed:
chore(remappings.txt): rm obsolete remapping

gitmp01 committed:
chore(CctpMessage): rm unused functions

gitmp01 committed:
chore(TokenBridgeCctp.t.sol): add coverage

gitmp01 committed:
fix: lint errors

gitmp01 committed:
chore(ccip-server): fix lint errors

gitmp01 committed:
chore(coverage.sh): ignore `TypedMemView` library

gitmp01 committed:
chore(eslint.config.mjs): ignore lib folder
